Cargo: decouple crate from main Cargo.toml workspace

phip1611 opened the pull request:
This is my solution for #7524 which we discussed in todays Cloud Hypervisor office hour. I think it improves certain things:

## TL;DR improvements

- more fearless refactorings: `cargo test|check|clippy` now **run on the whole workspace by default** (without explicit `--all/--workspace)`
- main `Cargo.toml` is not so overwhelming anymore
- We got rid of a lot of repetitions in CI files (no more `-D warnings` all the time)
- cleaned up technical debt of unused build.rs scripts

More info in ticket or commit message.

## Steps

- add new workspace member `cloud-hypervisor`
- move `./src` to new workspace member
- move `./tests` to new workspace member
- move relevant parts from Cargo.toml to new workspace member

phip1611 committed:
misc: improve test idempotence (unrelated improvement)

Signed-off-by: Philipp Schuster <philipp.schuster@cyberus-technology.de>
On-behalf-of: SAP philipp.schuster@sap.com

phip1611 committed:
build: remove useless build scripts (copy & paste artifact)

This will also prevent some useless rebuilds. Using `--verbose` we can
observe that the build.rs causes frequent useless rebuilds - having
less is a good thing. They come from the dependency of `build.rs` to
the local git repository.

Signed-off-by: Philipp Schuster <philipp.schuster@cyberus-technology.de>
On-behalf-of: SAP philipp.schuster@sap.com

phip1611 committed:
build: decouple Cargo.toml: split crate and workspace definitions

TL;DR: cargo clippy|check|... now runs on whole workspace by default.

## Steps

- add new workspace member `cloud-hypervisor`
- move `./src` to new workspace member
- move `./tests` to new workspace member
- move relevant parts from Cargo.toml to new workspace member
- kept necessary parts in main Cargo.toml, such as profile
  configurations

## About

The main Cargo.toml historically mixes workspace and crate definitions
for cloud-hypervisor and ch-remote. This makes it hard to read and
requires `--workspace` to run cargo clippy or cargo test on all
workspace members, which is counter-intuitive.

This patch separates the workspace from the crate definition in the main
Cargo.toml file. After this, cargo clippy, cargo test, etc., work on the
whole workspace naturally, giving a smoother developer experience. The
Cargo.toml without a package definition is also called a virtual
workspace or virtual manifest by Cargo [0].

Backporting is not a concern: CHV no longer backports, but the affected
files are rarely modified anyway.

[0] https://doc.rust-lang.org/cargo/reference/workspaces.html#virtual-workspace

Signed-off-by: Philipp Schuster <philipp.schuster@cyberus-technology.de>
On-behalf-of: SAP philipp.schuster@sap.com

phip1611 committed:
ci: replace `cargo rustc` -> `cargo build` to support virtual manifests

`cargo rustc` is incompatible with virtual manifests, so the CI needs to
 use cargo build instead. However, passing `RUSTFLAGS="-D warnings"` via
 the environment would propagate to all dependencies, and some of them
 currently fail to build under ``-D warnings` due to issues like [0]:

```
error: creating a mutable reference to mutable static
  --> src/temp.rs:97:5
   |
97 |     DIRS.pop()
   |     ^^^^^^^^^^ mutable reference to mutable static
```

To resolve this, apply ``-D warnings` only to the `cargo clippy`
commands (which apply to our workspace only) and avoid enforcing it for
the entire cargo build.

[0]: https://github.com/cloud-hypervisor/cloud-hypervisor/actions/runs/19962283528/job/57245376263?pr=7525

Signed-off-by: Philipp Schuster <philipp.schuster@cyberus-technology.de>
On-behalf-of: SAP philipp.schuster@sap.com

phip1611 committed:
test_infra: Adapt to workspace + make code more robust

Signed-off-by: Philipp Schuster <philipp.schuster@cyberus-technology.de>
On-behalf-of: SAP philipp.schuster@sap.com

phip1611 commented:
ping @RuoqingHe - I needed to take special care for the riscv job

https://github.com/cloud-hypervisor/cloud-hypervisor/pull/7525/commits/0259156f6154258e3b811792af56a8df8e198f4e

alyssais commented on CONTRIBUTING.md:
```diff
@@ -48,9 +48,9 @@ have specific features that you regularly develop against.
-cargo +nightly fmt -- --check || exit 1
-cargo check --locked --all --all-targets --tests || exit 1
-cargo clippy --locked --all --all-targets --tests -- -D warnings || exit 1
+cargo +nightly fmt --all -- --check || exit 1
```
The commit message for this only mentions removing `--all` — it doesn't explain why it's being added here.

alyssais reviewed (commented):
Like the direction. :)

phip1611 committed:
docs: adjust contributing.md: fix missing --all

Unrelated but necessary to also always format code for all
architectures and all features.

Signed-off-by: Philipp Schuster <philipp.schuster@cyberus-technology.de>
On-behalf-of: SAP philipp.schuster@sap.com

phip1611 committed:
docs: adjust contributing.md: remove need for --all (--workspace)

It is no longer needed to add `--all` (which is an alias for
`--workspace`). The documentation says "Commands run in the workspace
root will run against all workspace members by default" [0].

We however still need `--tests` as this activates the building of
tests in `<crate>/tests` directories.

[0] https://doc.rust-lang.org/cargo/reference/workspaces.html#virtual-workspace

Signed-off-by: Philipp Schuster <philipp.schuster@cyberus-technology.de>
On-behalf-of: SAP philipp.schuster@sap.com

alyssais commented on .github/workflows/build.yaml:
```diff
@@ -19,6 +19,8 @@ jobs:
           - x86_64-unknown-linux-gnu
           - x86_64-unknown-linux-musl
+    env:
+      RUSTFLAGS: -D warnings
```
Oh even worse, `RUSTC_WORKSPACE_WRAPPER` has to be a path, so we'd have to have a script like the following somewhere and set `RUSTC_WORKSPACE_WRAPPER` to the path to that:

```sh
#!/bin/sh
exec rustc -D warnings "$@"
```

phip1611 replied:
> and so I think we could solve this problem by catching warnings in the clippy checks rather than the build checks.

I love that, perfect! thanks

Fun fact for @rbradford @likebreath : this discussion actually took place during our in-person Cloud Hypervisor workshop at the @cyberus-technology office with @alyssais 🎉
